lib/SyTest/Federation/Client.pm

```diff
@@ -105,14 +124,78 @@ sub send_edu
          }
       ],
    );
+}
+
+sub send_event
+{
+   my $self = shift;
+   my %params = @_;
+
+   $self->send_transaction(
+      %params,
+      pdus => [ $params{event} ],
+   );
+}
+
+sub join_room
+{
+   my $self = shift;
+   my %args = @_;
+
+   my $server_name = $args{server_name};
+   my $room_id     = $args{room_id};
+   my $user_id     = $args{user_id};
+
+   my $store = $self->{datastore};
 
    $self->do_request_json(
-      method   => "PUT",
-      hostname => $params{destination},
-      uri      => "/send/$ts/",
+      method   => "GET",
+      hostname => $server_name,
+      uri      => "/make_join/$room_id/$user_id"
+   )->then( sub {
+      my ( $body ) = @_;
 
-      content => \%transaction,
-   )->then_done(); # response body is empty
+      my $protoevent = $body->{event};
+
+      my %member_event = (
+         ( map { $_ => $protoevent->{$_} } qw(
+            auth_events content depth prev_events prev_state room_id sender
+            state_key type ) ),
+
+         event_id         => $store->next_event_id,
+         origin           => $store->server_name,
+         origin_server_ts => $self->time_ms,
+      );
+
+      # TODO: really ought to sign it...
+
+      $self->do_request_json(
+         method   => "PUT",
+         hostname => $server_name,
+         uri      => "/send_join/$room_id/$member_event{event_id}",
+
+         content => \%member_event,
+      )->then( sub {
+         my ( $join_body ) = @_;
+         # SYN-490 workaround
+         $join_body = $join_body->[1] if ref $join_body eq "ARRAY";
+
+         my $room = SyTest::Federation::Room->new(
+            datastore => $store,
+            room_id   => $room_id,
+         );
+
+         my %done_event;
+         foreach my $event ( @{ $join_body->{auth_chain} }, @{ $join_body->{state} } ) {
+            $done_event{ $event->{event_id} }++ or
```

Comment:
Add each event only once. There'll possibly be duplicates between the auth chain and the current state, so this deduplicates them by event ID. Alternatively I suppose I could

    uniq_by { $_->{event_id} } @events...